### ISSUE TITLE
loader.js: `window.google` initialization conflict in presence of named google element 

### ISSUE THREAD

**Issue by aaemnnosttv** [google/google-visualization-issues#2936]
There appears to be a bug in https://www.gstatic.com/charts/loader.js.

When loading charts on a page that has an "named element" `google` (e.g. `<div id="google">`), this element will be referenced by default on the global `window` object as `window.google` so long that it isn't defined in JS yet as defined by the HTML spec: https://html.spec.whatwg.org/multipage/window-object.html#named-access-on-the-window-object. This isn't only about elements with such an `id` (although this is probably the most likely case), it can also happen with `name="google"` attribute but only for a handful of tags. 

When a DOM element is referenced by the `google` global, `loader.js` populates its properties onto the referenced instance of that `Element` rather than initializing a new object. At this point `window.google` will still has all of the expected methods on it even though it is also an `Element`. However, once `google.charts.load` is completed, the `window.google` global points to a new object, but **only contains the `visualization` property** while the others are left on the original `Element` which is no longer referenced by the `google` global.

This results in many properties to appear missing on the global `google` object once charts have been loaded, which can easily result in fatal errors if for example `google.charts.load()` is attempted to be called again as accessing a property of undefined.

I've created a repo to demonstrate this, which also includes a simple fix which can be applied in userland to prevent this from happening. 

https://aaemnnosttv.github.io/google-charts-loader-bug/ ([source](https://github.com/aaemnnosttv/google-charts-loader-bug))

TL;DR this can be prevented by initializing the `google` global before loading `loader.js`, but hopefully this can be fixed in the loader itself.

```html
<script>
if (window.google instanceof Element) {
	window.google = undefined;
}
</script>
<script src="https://www.gstatic.com/charts/loader.js"></script>
```

**Comment by dlaliberte** [comment]
Thanks, Evan, for investigating and reporting this issue.  It has likely been a problem since the beginning of Google Charts, and may well have caused confusing unresolved problems for at least a few people.  

The original reason that whatever is on window.google is preserved by the loader is that other APIs were also using the same symbol.  In particular, the old loader (loaded from www.google.com/jsapi) is on google.load, and many web pages still use that way of loading Google Charts.   All versions of Google Charts itself should also preserve whatever was on the global window.google symbol.

However, you are saying that after the new loader (loaded from www.gstatic.com/charts/loader.js) finishes loading the library packages, window.global is a new object.  This is unintentional, and may be due to changes in either the loader or the library.  As far as I know, the loader and the chart library itself were not changed for several versions (i.e. several years) regarding this aspect of how loading is done.  But I can imagine that refactorings to migrate the loader and library to es6 and TypeScript may have introduced this change.

If this problem only occurs when there is a preexisting window.google object which is an Element, then your workaround seems reasonable, and I am not so concerned about this problem.  I believe I can add the same workaround to the loader itself, though doing so could affect pages that depend on window.google continuing to be an Element.

But if this problem occurs regardless, that once the loader has loaded some packages of the library, the global window.google always no longer contains what it used to and it then points to a new object, that would mean that you could no longer call google.charts.load(), as you identified.    Could you clarify whether this problem occurs if there is no prior DOM element on window.google?

Also, which version of Google Charts are you using?  If the same problem applies to all previous versions, that makes more sense to me.  But if it is not a problem in older versions, then we have a clue to figure out what happened.

**Comment by aaemnnosttv** [comment]
Thanks for the quick response Daniel!

> The original reason that whatever is on window.google is preserved by the loader is that other APIs were also using the same symbol. In particular, the old loader (loaded from www.google.com/jsapi) is on google.load, and many web pages still use that way of loading Google Charts. All versions of Google Charts itself should also preserve whatever was on the global window.google symbol.

That makes sense, I know the global is shared by some other libraries like maps so I figured that was the intention. I haven't tested any other libraries to see if the loader has the same behavior.

> If this problem only occurs when there is a preexisting window.google object which is an Element, then your workaround seems reasonable, and I am not so concerned about this problem.

That seems to be the case, yes. If there is no conflicting element involved and `google` is defined as an object, any additional properties will be preserved as expected. Predefined falsy values are replaced with a new object, but the loader will fail if the global is a truthy value that can't receive new properties (like a string).

> I believe I can add the same workaround to the loader itself, though doing so could affect pages that depend on window.google continuing to be an Element.

True, although I don't see why any pages would particularly depend on it being an _Element_ specifically but I can imagine some sites may have code that simply checks `window.global` before setting something else on that object assuming that a library is loaded.

From that perspective, maybe the safer alternative would be to initialize the global using the current value if it's an object? This would preserve the element as the `google` object going forward (this might have other weird consequences though). i.e.

![image](https://user-images.githubusercontent.com/1621608/141698849-ea064998-68ba-48c8-89ce-c4532719c4f9.png)

> Could you clarify whether this problem occurs if there is no prior DOM element on window.google?

I'm not exactly sure what you mean. I'm happy to test something out, but it sounds like you're describing a more standard case where there wouldn't be a conflict. The problem is really about how the element is getting referenced on the same name on the window, but more like a fallback; it's not really defined (you can see this above). Once the `window.google` property is defined – to anything at all – the defined property will take precedence.

> Also, which version of Google Charts are you using? If the same problem applies to all previous versions, that makes more sense to me. But if it is not a problem in older versions, then we have a clue to figure out what happened.

I've tested it with every frozen version available (including `upcoming`) using `google.charts.load` and the behavior is the same, which leads me to believe it's the loader itself.

**Comment by dlaliberte** [comment]
I'll make a change to the loader to use something like your "safer alternative" workaround.   

I'm also wondering if this behavior (when window.google is an Element) is a browser-specific thing, or perhaps it is a safety feature of how top-level modules are created now, such that it avoids stepping on an existing DOM element, though it will reuse an existing object that is not a DOM element.

You may be aware that www.gstatic.com/charts/loader.js is the "top-level" loader, which then loads a "version-specific" loader that is different for every version of the library, except the pre-45 versions.   And the version-specific loader then loads the packages required.  The prior versions of the top-level loader are not available, though I could try to rebuild one if necessary, except that changes in the compiler and build environment could still result in differences.  Just FYI.

**Comment by aaemnnosttv** [comment]
> I'm also wondering if this behavior (when window.google is an Element) is a browser-specific thing, or perhaps it is a safety feature of how top-level modules are created now, such that it avoids stepping on an existing DOM element, though it will reuse an existing object that is not a DOM element.

I don't think it's browser-specific, unless a browser were implementing the spec incorrectly 😄  It's not a safety feature though as the HTML standard [warns about referencing elements in this way](https://html.spec.whatwg.org/multipage/window-object.html#named-access-on-the-window-object):

> As a general rule, relying on this will lead to brittle code. Which IDs end up mapping to this API can vary over time, as new features are added to the web platform, for example. Instead of this, use `document.getElementById()` or `document.querySelector()`.

> You may be aware that www.gstatic.com/charts/loader.js is the "top-level" loader, which then loads a "version-specific" loader that is different for every version of the library, except the pre-45 versions. And the version-specific loader then loads the packages required.

If I block the request for the version-specific loader, the problem doesn't seem to happen yet (`window.google` continues to reference the `Element` with the expected additional properties ([screenshot](https://user-images.githubusercontent.com/1621608/141836689-224f22b2-dfdb-4036-af80-14bd89cbc6f8.png))). Of course charts don't load (and the `onLoadCallback` is never called) so I'm not sure if that implies the bug is in the version-specific loader or if it's in the top-level loader and it was interrupted from getting that far.
